File: src/assembler/codex-agents-md-assembler.ts
/**
 * CodexAgentsMdAssembler — generates .codex/AGENTS.md from Nunjucks template.
 *
 * Operates in 3 phases:
 * 1. Collect extended context (scan agents, skills, hooks from .claude/)
 * 2. Build rendering context (24 flat config fields + extended fields)
 * 3. Render template and write output to .codex/AGENTS.md
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import yaml from "js-yaml";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import { buildDefaultContext } from "../template-engine.js";
import { resolveStack } from "../domain/resolver.js";
import type { AssembleResult } from "./rules-assembler.js";
import {
  DEFAULT_MODEL,
  SANDBOX_WORKSPACE_WRITE,
  isAccessibleDirectory,
  detectHooks,
  deriveApprovalPolicy,
  mapMcpServers,
} from "./codex-shared.js";

/** Metadata extracted from a generated agent .md file. */
export interface AgentInfo {
  readonly name: string;
  readonly description: string;
}

/** Metadata extracted from a generated skill's SKILL.md frontmatter. */
export interface SkillInfo {
  readonly name: string;
  readonly description: string;
  readonly userInvocable: boolean;
}

const TEMPLATE_PATH = "codex-templates/agents-md.md.njk";

/**
 * Scan a directory for agent .md files.
 *
 * @param agentsDir - Absolute path to the agents directory.
 * @returns Sorted array of AgentInfo with name and description.
 */
export function scanAgents(agentsDir: string): AgentInfo[] {
  if (!isAccessibleDirectory(agentsDir)) return [];
  const files = fs.readdirSync(agentsDir)
    .filter((f) => f.endsWith(".md"))
    .sort();
  const agents: AgentInfo[] = [];
  for (const file of files) {
    const name = file.replace(/\.md$/, "");
    const content = fs.readFileSync(
      path.join(agentsDir, file), "utf-8",
    );
    const description = extractDescription(content);
    agents.push({ name, description });
  }
  return agents;
}

/** Extract description from the first meaningful line of content. */
function extractDescription(content: string): string {
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed === "") continue;
    if (trimmed.startsWith("# ")) {
      return trimmed.slice(2).trim();
    }
    return trimmed;
  }
  return "";
}

/**
 * Scan a directory for skill subdirs containing SKILL.md.
 *
 * @param skillsDir - Absolute path to the skills directory.
 * @returns Sorted array of SkillInfo with name, description, userInvocable.
 */
export function scanSkills(skillsDir: string): SkillInfo[] {
  if (!isAccessibleDirectory(skillsDir)) return [];
  const entries = fs.readdirSync(skillsDir, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .sort((a, b) => a.name.localeCompare(b.name));
  const skills: SkillInfo[] = [];
  for (const entry of entries) {
    const skillMdPath = path.join(
      skillsDir, entry.name, "SKILL.md",
    );
    if (!fs.existsSync(skillMdPath)) continue;
    const content = fs.readFileSync(skillMdPath, "utf-8");
    skills.push(parseSkillFrontmatter(content, entry.name));
  }
  return skills;
}

/** Extract raw YAML frontmatter block between --- delimiters. */
function extractFrontmatterBlock(content: string): string | null {
  const lines = content.split("\n");
  if (lines.length === 0 || lines[0]!.trim() !== "---") return null;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i]!.trim() === "---") {
      return lines.slice(1, i).join("\n");
    }
  }
  return null;
}

/** Parse YAML frontmatter from SKILL.md content using js-yaml. */
function parseSkillFrontmatter(
  content: string,
  dirName: string,
): SkillInfo {
  const block = extractFrontmatterBlock(content);
  if (!block) {
    return { name: dirName, description: "", userInvocable: true };
  }
  const parsed = yaml.load(block) as Record<string, unknown> | null;
  if (!parsed || typeof parsed !== "object") {
    return { name: dirName, description: "", userInvocable: true };
  }
  const name = typeof parsed["name"] === "string"
    ? parsed["name"] : dirName;
  const rawDesc = parsed["description"];
  const description = typeof rawDesc === "string"
    ? rawDesc.trim() : "";
  const userInvocable = parsed["user-invocable"] !== false;
  return { name, description, userInvocable };
}

/**
 * Build extended context for AGENTS.md template rendering.
 *
 * Merges the 24 flat config fields with extended fields:
 * resolved_stack, agents_list, skills_list, has_hooks,
 * mcp_servers, security_frameworks, observability,
 * model, approval_policy, sandbox_mode.
 */
export function buildExtendedContext(
  config: ProjectConfig,
  agents: AgentInfo[],
  skills: SkillInfo[],
  hasHooks: boolean,
): Record<string, unknown> {
  const resolved = resolveStack(config);
  return {
    ...buildDefaultContext(config),
    observability: config.infrastructure.observability.tool,
    resolved_stack: {
      buildCmd: resolved.buildCmd,
      testCmd: resolved.testCmd,
      compileCmd: resolved.compileCmd,
      coverageCmd: resolved.coverageCmd,
    },
    agents_list: agents,
    skills_list: skills.map(toTemplateSkill),
    has_hooks: hasHooks,
    mcp_servers: mapMcpServers(config),
    security_frameworks: [...config.security.frameworks],
    model: DEFAULT_MODEL,
    approval_policy: deriveApprovalPolicy(hasHooks),
    sandbox_mode: SANDBOX_WORKSPACE_WRITE,
  };
}

/** Map SkillInfo (camelCase) to template context (snake_case). */
function toTemplateSkill(
  s: SkillInfo,
): { name: string; description: string; user_invocable: boolean } {
  return {
    name: s.name,
    description: s.description,
    user_invocable: s.userInvocable,
  };
}

/** Collect agents, skills, and hooks from .claude/ output directory. */
function collectContext(
  claudeDir: string,
  warnings: string[],
): { agents: AgentInfo[]; skills: SkillInfo[]; hasHooks: boolean } {
  const agents = scanAgents(path.join(claudeDir, "agents"));
  const skills = scanSkills(path.join(claudeDir, "skills"));
  const hooksDir = path.join(claudeDir, "hooks");
  const hasHooks = detectHooks(hooksDir);
  if (agents.length === 0) {
    warnings.push("No agents found in output directory");
  }
  if (skills.length === 0) {
    warnings.push("No skills found in output directory");
  }
  return { agents, skills, hasHooks };
}

/** Render template and write AGENTS.md to outputDir. */
function renderAndWrite(
  engine: TemplateEngine,
  context: Record<string, unknown>,
  outputDir: string,
  warnings: string[],
): AssembleResult {
  let rendered: string;
  try {
    rendered = engine.renderTemplate(TEMPLATE_PATH, context);
  } catch (error: unknown) {
    const message = error instanceof Error
      ? error.message
      : String(error);
    if (message.includes("not found") || message.includes("ENOENT")) {
      warnings.push(`Template not found: ${TEMPLATE_PATH}`);
      return { files: [], warnings };
    }
    throw error;
  }
  fs.mkdirSync(outputDir, { recursive: true });
  const dest = path.join(outputDir, "AGENTS.md");
  fs.writeFileSync(dest, rendered, "utf-8");
  return { files: [dest], warnings };
}

/** Generates .codex/AGENTS.md from Nunjucks template. */
export class CodexAgentsMdAssembler {
  /** Generate .codex/AGENTS.md by scanning .claude/ output and rendering template. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    _resourcesDir: string,
    engine: TemplateEngine,
  ): AssembleResult {
    const warnings: string[] = [];
    const claudeDir = path.join(path.dirname(outputDir), ".claude");
    const { agents, skills, hasHooks } = collectContext(
      claudeDir, warnings,
    );
    const context = buildExtendedContext(
      config, agents, skills, hasHooks,
    );
    return renderAndWrite(engine, context, outputDir, warnings);
  }
}

File: java/src/main/java/dev/iadev/assembler/CodexAgentsMdAssembler.java
package dev.iadev.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.stack.StackResolver;
import dev.iadev.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;
import org.yaml.snakeyaml.LoaderOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.SafeConstructor;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Generates {@code AGENTS.md} at the project root for
 * OpenAI Codex CLI.
 *
 * <p>Operates in 3 phases:
 * <ol>
 *   <li>Collect extended context by scanning agents, skills,
 *       and hooks from the {@code .claude/} output</li>
 *   <li>Build rendering context merging the 25 flat config
 *       fields with extended fields</li>
 *   <li>Render the Pebble template and write
 *       {@code AGENTS.md}</li>
 * </ol>
 *
 * <p>This is the seventeenth assembler in the pipeline
 * (position 17 of 23 per RULE-005). Its target is
 * {@link AssemblerTarget#ROOT}.</p>
 *
 * @see Assembler
 * @see CodexShared
 */
public final class CodexAgentsMdAssembler
        implements Assembler {

    private static final String TEMPLATE_PATH =
            "codex-templates/agents-md.md.njk";

    /**
     * {@inheritDoc}
     *
     * <p>Generates {@code AGENTS.md} by scanning
     * {@code .claude/} output and rendering the codex
     * agents template.</p>
     */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        List<String> warnings = new ArrayList<>();

        Path claudeDir = outputDir.getParent()
                .resolve(".claude");

        List<AgentInfo> agents = scanAgents(
                claudeDir.resolve("agents"));
        List<SkillInfo> skills = scanSkills(
                claudeDir.resolve("skills"));
        HookPresence hookPresence = HookPresence.of(
                CodexShared.detectHooks(
                        claudeDir.resolve("hooks")));

        if (agents.isEmpty()) {
            warnings.add(
                    "No agents found in output directory");
        }
        if (skills.isEmpty()) {
            warnings.add(
                    "No skills found in output directory");
        }

        Map<String, Object> context =
                buildExtendedContext(
                        config, agents, skills,
                        hookPresence);

        String rendered = engine.render(
                TEMPLATE_PATH, context);

        CopyHelpers.ensureDirectory(outputDir);
        Path dest = outputDir.resolve("AGENTS.md");
        CopyHelpers.writeFile(dest, rendered);

        List<String> result = new ArrayList<>();
        result.add(dest.toString());
        return result;
    }

    /**
     * Scans a directory for agent {@code .md} files.
     *
     * @param agentsDir the agents directory path
     * @return sorted list of {@link AgentInfo}
     */
    static List<AgentInfo> scanAgents(Path agentsDir) {
        if (!CodexShared.isAccessibleDirectory(agentsDir)) {
            return List.of();
        }
        List<AgentInfo> agents = new ArrayList<>();
        try (DirectoryStream<Path> stream =
                     Files.newDirectoryStream(
                             agentsDir, "*.md")) {
            List<Path> sorted = new ArrayList<>();
            stream.forEach(sorted::add);
            sorted.sort((a, b) -> a.getFileName().toString()
                    .compareTo(b.getFileName().toString()));
            for (Path file : sorted) {
                String name = file.getFileName().toString()
                        .replaceFirst("\\.md$", "");
                String content = Files.readString(
                        file, StandardCharsets.UTF_8);
                String description =
                        extractDescription(content);
                agents.add(
                        new AgentInfo(name, description));
            }
        } catch (IOException e) {
            return List.of();
        }
        return agents;
    }

    /**
     * Extracts description from the first meaningful line
     * of content.
     *
     * @param content the file content
     * @return the description string
     */
    static String extractDescription(String content) {
        for (String line : content.split("\n")) {
            String trimmed = line.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            if (trimmed.startsWith("# ")) {
                return trimmed.substring(2).trim();
            }
            return trimmed;
        }
        return "";
    }

    /**
     * Scans a directory for skill subdirs containing
     * {@code SKILL.md}.
     *
     * @param skillsDir the skills directory path
     * @return sorted list of {@link SkillInfo}
     */
    static List<SkillInfo> scanSkills(Path skillsDir) {
        if (!CodexShared.isAccessibleDirectory(skillsDir)) {
            return List.of();
        }
        List<SkillInfo> skills = new ArrayList<>();
        try (DirectoryStream<Path> stream =
                     Files.newDirectoryStream(skillsDir)) {
            List<Path> sorted = new ArrayList<>();
            stream.forEach(sorted::add);
            sorted.sort((a, b) -> a.getFileName().toString()
                    .compareTo(b.getFileName().toString()));
            for (Path entry : sorted) {
                if (!Files.isDirectory(entry)) {
                    continue;
                }
                Path skillMd =
                        entry.resolve("SKILL.md");
                if (!Files.exists(skillMd)) {
                    continue;
                }
                String content = Files.readString(
                        skillMd, StandardCharsets.UTF_8);
                skills.add(parseSkillFrontmatter(
                        content,
                        entry.getFileName().toString()));
            }
        } catch (IOException e) {
            return List.of();
        }
        return skills;
    }

    /**
     * Extracts raw YAML frontmatter block between
     * {@code ---} delimiters.
     *
     * @param content the file content
     * @return the YAML block or null if not found
     */
    static String extractFrontmatterBlock(String content) {
        String[] lines = content.split("\n");
        if (lines.length == 0
                || !"---".equals(lines[0].trim())) {
            return null;
        }
        for (int i = 1; i < lines.length; i++) {
            if ("---".equals(lines[i].trim())) {
                StringBuilder sb = new StringBuilder();
                for (int j = 1; j < i; j++) {
                    if (j > 1) {
                        sb.append("\n");
                    }
                    sb.append(lines[j]);
                }
                return sb.toString();
            }
        }
        return null;
    }

    /**
     * Parses YAML frontmatter from {@code SKILL.md} content
     * using SnakeYAML.
     *
     * @param content the SKILL.md file content
     * @param dirName the directory name as fallback
     * @return a {@link SkillInfo} with extracted metadata
     */
    @SuppressWarnings("unchecked")
    static SkillInfo parseSkillFrontmatter(
            String content, String dirName) {
        String block = extractFrontmatterBlock(content);
        if (block == null) {
            return new SkillInfo(dirName, "", true);
        }
        Yaml yaml = new Yaml(new SafeConstructor(
                new LoaderOptions()));
        Object parsed = yaml.load(block);
        if (!(parsed instanceof Map)) {
            return new SkillInfo(dirName, "", true);
        }
        Map<String, Object> map =
                (Map<String, Object>) parsed;

        String name = map.get("name") instanceof String
                ? (String) map.get("name") : dirName;
        Object rawDesc = map.get("description");
        String description = rawDesc instanceof String
                ? ((String) rawDesc).trim() : "";
        Object rawInvocable = map.get("user-invocable");
        boolean userInvocable =
                !Boolean.FALSE.equals(rawInvocable);
        return new SkillInfo(name, description, userInvocable);
    }

    /**
     * Builds extended context for {@code AGENTS.md} template
     * rendering.
     *
     * <p>Merges the 25 flat config fields with extended
     * fields: resolved_stack, agents_list, skills_list,
     * has_hooks, mcp_servers, security_frameworks,
     * observability, model, approval_policy,
     * sandbox_mode.</p>
     *
     * @param config       the project configuration
     * @param agents       the scanned agents
     * @param skills       the scanned skills
     * @param hookPresence whether hooks were detected
     * @return the template context map
     */
    static Map<String, Object> buildExtendedContext(
            ProjectConfig config,
            List<AgentInfo> agents,
            List<SkillInfo> skills,
            HookPresence hookPresence) {
        var resolved = StackResolver.resolve(config);
        Map<String, Object> ctx =
                new LinkedHashMap<>(
                        ContextBuilder.buildContext(config));

        ctx.put("observability",
                config.observabilityTool());

        Map<String, String> resolvedStack =
                new LinkedHashMap<>();
        resolvedStack.put("buildCmd", resolved.buildCmd());
        resolvedStack.put("testCmd", resolved.testCmd());
        resolvedStack.put("compileCmd",
                resolved.compileCmd());
        resolvedStack.put("coverageCmd",
                resolved.coverageCmd());
        ctx.put("resolved_stack", resolvedStack);

        ctx.put("agents_list", agents);

        List<Map<String, Object>> templateSkills =
                new ArrayList<>();
        for (SkillInfo skill : skills) {
            Map<String, Object> entry =
                    new LinkedHashMap<>();
            entry.put("name", skill.name());
            entry.put("description", skill.description());
            entry.put("user_invocable",
                    skill.userInvocable());
            templateSkills.add(entry);
        }
        ctx.put("skills_list", templateSkills);

        ctx.put("has_hooks", hookPresence.hasHooks());
        ctx.put("mcp_servers",
                CodexShared.mapMcpServers(config));
        ctx.put("security_frameworks",
                new ArrayList<>(
                        config.security().frameworks()));
        ctx.put("model", CodexShared.DEFAULT_MODEL);
        ctx.put("approval_policy",
                CodexShared.deriveApprovalPolicy(
                        hookPresence));
        ctx.put("sandbox_mode",
                CodexShared.SANDBOX_WORKSPACE_WRITE);

        return ctx;
    }

}
